### ISSUE TITLE
Restore could handle correct importing directly from Route53 api formatting

### ISSUE THREAD

**Issue by gravyboat** [swelljoe/r53z#1]
I've got a script that's backing up Route53 domains in their typical JSON format. Right now you can't use this tool to import these domains since the format is modified after domains are pulled down from the API and it expects the same on restore. It would be nice if the `restore` function could support the standard API format, primarily this is an issue for items such as:

API Format:

```
    {
        "Name": "yep.domain.com.",
        "ResourceRecords": [
            {
                "Value": "\"fjskdf sdlkf dsklfds j\""
            },
            {
                "Value": "\"something else\""
            }
        ],
        "TTL": 300,
        "Type": "TXT"
    }

```

r53z format:

```
  {
    "name": "yep.domain.com.",
    "type": "TXT",
    "ttl": 300,
    "resource_records": [
      {
        "value": "\"fjskdf sdlkf dsklfds j\""
      },
      {
        "value": "\"something else\""
      }
    ]
  }
```

If you try to restore the API zone you'll see an error similar to this:

```
parameter validator found 6 errors:
  - missing required parameter params[:change_batch][:changes][0][:resource_record_set][:name]
  - missing required parameter params[:change_batch][:changes][0][:resource_record_set][:type]
  - unexpected value at params[:change_batch][:changes][0][:resource_record_set][:Name]
  - unexpected value at params[:change_batch][:changes][0][:resource_record_set][:Type]
  - unexpected value at params[:change_batch][:changes][0][:resource_record_set][:ResourceRecords]
  - unexpected value at params[:change_batch][:changes][0][:resource_record_set][:AliasTarget]
```

So it mostly seems to be the conversion from camel case to underscores.

**Comment by gravyboat** [comment]
The other option I've been playing with a bit is a `convert` function, but I don't like that anywhere near as much as just integrating it.

**Comment by swelljoe** [comment]
Yeah, that's do-able. The data _is_ a bit shady, here, since it gets shipped back and forth between Ruby structs and JSON objects (and JSON doesn't know about Ruby symbols, etc.), and there's some lossiness happening. I'm not actually sure where the camel case to unders conversion is coming from, though; I didn't explicitly tell it to do that...it seems to be coming in automatically from the Ruby JSON encode, or maybe even from the Ruby version of the API.

Anyway, I'd noticed this mismatch, too, when I was first pulling in and out zones for testing with another script I'd written in Python. So, yeah, I can make it work with either (or maybe just make it always export the "right" way and work with that). Will probably be Thursday or Friday, as I'm rolling out a new release of Virtualmin.

**Comment by gravyboat** [comment]
I actually noticed the same thing you did (regarding the camel case conversion) earlier today when I was trying to modify the code locally to do so. I was going to prep a PR for it but couldn't figure out exactly what was causing it. No rush on the implementation, it would just be cool to have. Right now cli53 (https://github.com/barnybug/cli53#getting-started) only handles text, so even having it in JSON is nice. If it was in the same format though that would really give this a leg up in terms of ease of use.

**Comment by gravyboat** [comment]
Alright I did some research this morning and it's definitely due to the way the aws ruby sdk works (https://gitter.im/aws/aws-sdk-ruby/archives/2015/03/05). Something like the plissken gem (https://github.com/futurechimp/plissken) could address this. The only thing to make sure of is whether this works correctly on string objects (or that the file contents get imported and converted first).

**Comment by gravyboat** [comment]
I worked on this for a little while earlier and honestly it's a real pain in the butt. We have to convert from an array to a hash, then back to an array to 'prettify' the JSON, otherwise you get the hash keys in the array so the format is messed up when you write the file. I'm not sure what the best way to get around this is, it seems a bit problematic.

**Comment by swelljoe** [comment]
I poked around a bit in the aws-sdk today, and found how all this works; I was hoping for a simple spot where I could add a constructor option to disable it, but the internals rely heavily on these objects being normalized to underscores, so lots of internal functions are expecting snake case.

There's an issue about it here: https://github.com/aws/aws-sdk-ruby/issues/970

At least, I assume that's the format you're talking about? (e.g. generated by the Amazon AWS cli)

Interestingly, the comments there indicate that even the AWS cli is not producing "native" output from the API endpoints, and is also munging it. So, everybody has their own representation, the Ruby one just happens to be the most divergent (well, cli53 is even further out with BIND zone file output, which they put a lot of effort into).

While a generic solution that would work for all kinds of AWS requests seems quite difficult, a solution specific to our use case is not so scary, I don't think. I'm going to try adding a simple mapping layer when the other format is detected; i.e. just looping over all of the keys and running the aws underscore function (in seashorse/util.rb) on it. It'll be a little clumsy, but I think it's workable. There's already a bit of munging happening between dumping and restoring (because of differences in what we get vs. what we have to send), so this probably isn't too clunky.

Also, speaking of cli53, I had a look at their code back in the beginning of all of this. Adding JSON zone output would not be difficult (I even have a stubbed out version somewhere around here, that doesn't actually work yet, but I'd probably be able to make to the right thing in an hour or two), but it doesn't deal with a bunch of data that I think is necessary for your use case, so getting it to dump JSON zones would only be a small piece of the puzzle. Though, I suspect it'd be easier to make r53z do everything cli53 does (except BIND format output because that's a silly misfeature), and do it with the same command line options. Go is a pretty verbose language; I'm surprised at how much code it takes cli53 to do what it does.

Anyway, I'm gonna poke at conversion of cli JSON as soon as I can get over to my co-working space tonight; my internet here is slow enough to make working tiring.

**Comment by gravyboat** [comment]
Sounds good, thanks for looking at this @swelljoe. This week I put together a little lambda backup script (via the CLI so it uses their format) that gets the output (I believe) identical to what r53z handles, it's just in standard json format so it's not pretty formatted and obviously it uses camel case. If you want to play with it I've included it below. It's written in nodejs so if you don't want to mess with that nonsense I can see about getting an example zone file on Monday:

```
// Note that this script is triggered via a cloudwatch events job that is monitoring cloudtrail
const AWS = require("aws-sdk");
const zlib = require("zlib");

function doUpload(err, data) {
  const S3 = new AWS.S3();

  if (err) {
    console.log(JSON.stringify(err));
    return;
  }

  const params = {
    Body:   JSON.stringify(data),
    Bucket: "your-bucket",
    Key:    (data.ResourceRecordSets) ? `zones/${data.ResourceRecordSets[0].Name}json` : `zones/${data.HostedZone.Name}zoneinfo.json`
  }

  S3.upload(params, function(err, data) {
    console.log(JSON.stringify(err));
    return;
  });
}

exports.handle = function(event, context) {
  const Route53 = new AWS.Route53();
  var payload = new Buffer(event.awslogs.data, "base64");
  zlib.gunzip(payload, function(e, result) {
    if (e) {
      context.fail(e);
    }
    else {
      result = JSON.parse(result.toString("utf8"));
      var domainId = "";
      result.logEvents.forEach(function(entry) {
        if (entry.hasOwnProperty("message")) {
          console.log("message found!");
          convertEntryString = JSON.stringify(entry);
          convertEntryJson = JSON.parse(convertEntryString);
          jsonMessage = JSON.parse(convertEntryJson.message);
          console.log(jsonMessage);
          domainId = jsonMessage.requestParameters.hostedZoneId ||
                     jsonMessage.responseElements.hostedZone.id ||
                     jsonMessage.requestParameters.id;
          console.log(domainId);

          if (zoneId === "undefined" || zoneId !== domainId) {
            var zoneId  = jsonMessage.requestParameters.hostedZoneId ||
                          jsonMessage.responseElements.hostedZone.id ||
                          jsonMessage.requestParameters.id;
            Route53.getHostedZone({ Id: zoneId }, doUpload);
            Route53.listResourceRecordSets({ HostedZoneId: zoneId }, doUpload);
          }
        }
      });
    }
  });
}
```

I'm honestly surprised by how much work it took to get this output formatted and into a usable state, they definitely didn't make it easy.

**Comment by swelljoe** [comment]
I've checked in a couple of changes that allow it to read a record-set file coming out of the aws cli (and I think put it into a usable data structure, though I haven't written tests to prove that yet). I'm stuck on Windows at the moment, due to a tea spillage incident, so I don't have node.js setup (and I'm kinda Windows-stupid, so it always takes me longer than expected to make anything go on Windows). I doubt it actually works for restores or anything useful just yet, but the camel-case to underscore issue is seemingly resolved, at least for record sets.

I'll get some tests written, and will see about getting some actual zone files out of your node app. I'd like this to work with my Perl and Python scripts, too, which are all spitting out something closer to what aws cli is spitting out, instead of what we get from the ruby API. I'll probably get around to making it write them in this format, as well.

**Comment by swelljoe** [comment]
Ok, so restores are working now from exports created with the aws cli tool.

e.g. you can do the following (you'll have to lookup the id of your zone since the aws cli doesn't work with names very well):

```
$ aws route53 get-hosted-zone --id LJHDHJ37JGJJK > dumps/zonename.com.zoneinfo.json
$ aws route53 list-resource-record-sets --hosted-zone-id LJHDHJ37JGJJK > dumps/zonename.com.json
```

And, then you can restore that zone using r53z in the usual manner. At least, I've been able to do this just now. I still haven't written tests, as I was just tinkering with whether the right data could be pulled out in a reasonable fashion from the aws cli. It may also be clunky on delegation sets (none of my regular zones have a delegation set ID, so it is untested until tomorrow).

I'm expecting to get Linux running on my new lappy either Monday or Tuesday, so I'll be better equipped to tinker with node stuff, so I can see what format your script is spitting out. But, if it's the same output as the aws cli and the Python SDK, the version in git now should do the trick.

I'll probably eventually get around to making it possible to specify the metadata currently provided by the zoneinfo file on the command line, so only a resource records set is required to restore (as long as you specify the name, delegation set ID, and comment).

**Comment by gravyboat** [comment]
Sounds good. I'll do some testing on this today to confirm that it works correctly. If for some reason the output from my script isn't matching I'll paste an example in here and we can compare notes, then I'll update my script to match the expected default output.

**Comment by gravyboat** [comment]
Okay, did some quick testing this morning, and here's where the differences are (note I ran this through Python's `json.tool` so it looked better, by default mine is 'saving' space so it isn't printed nicely and is all on one line). The record details I'm pulling have a couple additional items which don't seem to belong:

```
{
    "IsTruncated": false,
    "MaxItems": "100",
    "ResourceRecordSets": [
        {
            "AliasTarget": {
                "DNSName": "some-elb.elb.amazonaws.com.",
                "EvaluateTargetHealth": false,
                "HostedZoneId": "ASDF1234FAKEFAKEFAKE"
            },
            "Name": "domain.com.com.",
            "ResourceRecords": [], 
            "Type": "A" 
        },
```

versus:

```
{
    "ResourceRecordSets": [
        {
            "AliasTarget": {
                "DNSName": "some-elb.elb.amazonaws.com.",
                "EvaluateTargetHealth": false,
                "HostedZoneId": "ASDF1234FAKEFAKEFAKE"
            },
            "Name": "domain.com.",
            "Type": "A" 
        },
```

So for some reason the lookup method being used via the node script in lambda includes a few extra items, these exist at the end and should be relatively easy to drop off, for the ResourceRecords value I'm not exactly sure why that's getting dropped in there, but I have to imagine it has to do with AWS not wanting this type of requested zone to be used for backups (it specifically says not to use it for importing). For `zoneinfo files` the differences are pretty minimal and exist at the end of the file:

```
    },  
    "VPCs": []
}
```

versus:

```
    }   
}
```

So overall pretty minor issues. I just need to look at some nodejs functionality in terms of dropping things to see if it's possible to easily pop those few items out.

Edit: I do find it odd that these extra items are somehow getting added specifically for the nodejs SDK and not the Python SDK. Seems very odd.

**Comment by swelljoe** [comment]
Hmm...It should actually already be handling what you pasted. There's a bit in the read function that removes the outer layer of the data structure if it exists and just returns the ResourceRecordSets array, which should make the import happy. So only things within ResourceRecordSets will cause problems (I think). 

The only thing I see different in the two pasted bits is the ResourceRecords array, which being empty shouldn't cause problems. But, maybe it does. Can you include an example of the full output of your script? If I can run it through here, I can sort out the problem in no time, I'm sure.

**Comment by gravyboat** [comment]
As soon as I get logged in tomorrow morning I'll get you the `json` and `zonefile.json` output for some test domains. It's going to look like crap because of how the return is handled (single line json for minimal storage sizes) but it should be okay. Do you want any specific options added on to the ruby script invocation itself to get more debugging?

**Comment by gravyboat** [comment]
Alright sorry about the delay this morning. Here's an example `zone`:

```
{"ResourceRecordSets":[{"Name":"test.com.","Type":"A","ResourceRecords":[],"AliasTarget":{"HostedZoneId":"FAK3Z0NE","DNSName":"some-elb.us-east-1.elb.amazonaws.com.","EvaluateTargetHealth":false}},{"Name":"test.com.","Type":"NS","TTL":172800,"ResourceRecords":[{"Value":"ns-479.awsdns-59.com."},{"Value":"ns-1192.awsdns-21.org."},{"Value":"ns-559.awsdns-05.net."},{"Value":"ns-1916.awsdns-47.co.uk."}]},{"Name":"test.com.","Type":"SOA","TTL":900,"ResourceRecords":[{"Value":"ns-479.awsdns-59.com. awsdns-hostmaster.amazon.com. 1 7200 900 1209600 86400"}]},{"Name":"test.com.","Type":"TXT","TTL":300,"ResourceRecords":[{"Value":"\"second text record\""},{"Value":"\"v=spf1 include:relay.some-fake-mail.net +a +mx +ip4:199.47.222.251 ?all\""},{"Value":"\"third text record\""}]},{"Name":"bleeblah.test.com.","Type":"A","TTL":300,"ResourceRecords":[{"Value":"5.4.3.2"}]},{"Name":"bleeblah.test.com.","Type":"MX","TTL":300,"ResourceRecords":[{"Value":"0 google.com."},{"Value":"5 a.google.com."}]},{"Name":"bleeblah.test.com.","Type":"TXT","TTL":300,"ResourceRecords":[{"Value":"\"fjskdf sdlkf dsklfds j\""},{"Value":"\"something else\""}]},{"Name":"sigh.test.com.","Type":"CNAME","TTL":300,"ResourceRecords":[{"Value":"google.com."}]},{"Name":"holysmokes.test.com.","Type":"TXT","TTL":300,"ResourceRecords":[{"Value":"\"test it\""}]},{"Name":"logloglog.test.com.","Type":"TXT","TTL":300,"ResourceRecords":[{"Value":"\"lets do it\""}]},{"Name":"route53backup.test.com.","Type":"TXT","TTL":300,"ResourceRecords":[{"Value":"\"route53 backup test\""}]},{"Name":"testing.test.com.","Type":"A","TTL":300,"ResourceRecords":[{"Value":"1.2.3.4"}]},{"Name":"testing.test.com.","Type":"MX","TTL":300,"ResourceRecords":[{"Value":"0 bleeblah.com."},{"Value":"5 wee.bleeblah.com."}]},{"Name":"testing.test.com.","Type":"TXT","TTL":300,"ResourceRecords":[{"Value":"\"this is a test\""},{"Value":"\"wooefoefejfefweofjwf ewf wekfjw fk\""}]},{"Name":"wooooo.test.com.","Type":"A","TTL":300,"ResourceRecords":[{"Value":"44.44.44.44"}]},{"Name":"wooooo.test.com.","Type":"MX","TTL":300,"ResourceRecords":[{"Value":"0 bla.testing.com."}]},{"Name":"www.test.com.","Type":"CNAME","TTL":3600,"ResourceRecords":[{"Value":"test.com."}]},{"Name":"yep.test.com.","Type":"A","TTL":300,"ResourceRecords":[{"Value":"12.123.12.12"}]},{"Name":"yep.test.com.","Type":"MX","TTL":300,"ResourceRecords":[{"Value":"0 google.com."},{"Value":"5 a.google.com."}]},{"Name":"yep.test.com.","Type":"TXT","TTL":300,"ResourceRecords":[{"Value":"\"fjskdf sdlkf dsklfds j\""},{"Value":"\"something else\""}]}],"IsTruncated":false,"MaxItems":"100"}
```

And here's an example `zoneinfo`:

```
{"HostedZone":{"Id":"/hostedzone/FAK3Z0NE","Name":"test.com.","CallerReference":"r53z-create-va1","Config":{"PrivateZone":false},"ResourceRecordSetCount":20},"DelegationSet":{"Id":"/delegationset/D3LS3T","CallerReference":"something","NameServers":["ns-479.awsdns-59.com","ns-1192.awsdns-21.org","ns-559.awsdns-05.net","ns-1916.awsdns-47.co.uk"]},"VPCs":[]}
```

**Comment by swelljoe** [comment]
So, aws-sdk is choking on the empty ResourceRecords field (in the first A record at the top of the zone records JSON); when given a Value, it makes it past that point and is able to create the zone. Does that zone item actually have an empty ResourceRecords array? I can't get it to create a record like that using the Ruby SDK, and that field is listed as required in the docs for the change_resource_record_set API call.

But, even with that fixed, I'm still getting an error "Aws::Route53::Errors::InvalidInput: Invalid request", so there's still something it doesn't like in there. But, it does create the zone, when I run the import in the REPL (which also gives me the AWS exceptions rather than catching it). Not sure if it gets all of its records, but it definitely gets most of them.

Anyway, just an update. I don't see why it's still spitting errors...it's not obvious to me what is invalid about it (aside from the empty ResourceRecord item), but I'll sort it out shortly, I'm sure, now that I'm able to reproduce the problem.

**Comment by gravyboat** [comment]
So I just double checked another (much simpler) domain that I have access to, and it also generates the empty `ResourceRecords` for the `A` record when using the nodejs sdk for aws. This leads me to believe it simply generates an empty `ResourceRecords` array for every entry in the zone file. If it doesn't have any actual records, then it just leaves it blank. Poor design for sure in terms of not standardizing it.

**Comment by swelljoe** [comment]
OK, so I think all of the issues I was seeing were in that AliasTarget section. When I get rid of it entirely, it successfully creates the zone with all records.

I don't have CloudFront or a load balancer setup on my account, currently, so it's non-trivial to test that locally (it won't let me create dummy entries with fake data).

In the short term I'll add a special case for empty ResourceRecords; it looks like either removing the line entirely or adding {Value: []} works to get around that empty list error. I need to test that on a record that isn't an AliasTarget record, so I can see it working to make sure my theory on that is right. It's plausible that the entirety of the problem you're seeing (since you presumably have valid AliasTarget info) is that one record that causes the Ruby API to choke. I might also do a bit of tinkering with a Python import of the same data. I'm curious if this is just a quirk of the Ruby API, or if it's a bug-like behavior of the JavaScript API that it'll export something with bogus data that can't be imported. I guess it'd also be useful to test importing with the JavaScript API, to see if it'll silently shove that record in there with an empty ResourceRecords list. It _looks_ like it is the endpoint doing the complaining, but I haven't actually dug in deep enough to know; the Ruby API does some of its own validation, as well.

I'll check in a version that does something with an empty list later today (once I get to the co-working space), and we can see if that maybe allows your records to import. You could also try with a test zone that does not have any AliasTarget sections (and no empty ResourceRecords lists) in the export. It is otherwise seemingly working for me, with the test zone you sent and the current git version of the app.

**Comment by swelljoe** [comment]
OK, Ruby SDK chokes no matter what I do with that empty ResourceRecords field. It behaved differently when it was part of the AliasTarget section (it still was failing but with different errors, so I figured setting it to an empty record or removing it was "fixing" it), but when given a zone with any of the options I came up with, it gives an error and fails to create that particular record; though the zone creation and other record creation succeeds.

So, I've re-factored the create function to allow an extra bit of munging to go on in between creation of the zone and creation of the records; this doesn't actually solve the problem of the Ruby SDK not allowing any kind of empty ResourceRecords there, under any circumstances, but it does allow me to create test cases that tinker with the records; we need a valid zone ID for an AliasTarget that points to a record within the same zone, so the two steps have to be separated out.

I don't think I'll finish that tonight, as it's more work than I expected to need to do (I thought I could just munge in an empty "Value": [] and be done with it, but that won't work), and I only set aside an hour or so for tinkering with this.

Quick question, if you know the answer: In that AliasTarget, if the ResourceRecords field were set to something nonsensical (like "Value": "127.0.0.1"), would that break things? It seems like it is ignored, but it is listed as required in the docs. I can't figure out what you're actually supposed to put there for something like an A record alias for a CloudFront, if not an empty list (but empty list is definitely rejected). This really seems like a weird implementation detail that just isn't documented very well. I guess it doesn't matter for some APIs as it seems to accept an empty list there, so maybe the Amazon folks didn't think it needed documenting any more clearly. I might need to ping the Ruby SDK folks for clarity.

**Comment by gravyboat** [comment]
What a pain, thanks for taking a look into that. I have no idea about the AliasTarget, but it definitely seems like it would cause an issue based on what it is supposed to represent, here's what th docs are saying:

```
Instead of an IP address or a domain name, an alias resource record set contains a pointer to a CloudFront distribution, an Elastic Beanstalk environment, an ELB load balancer, an Amazon S3 bucket that is configured as a static website, or another Amazon Route 53 resource record set in the same hosted zone.
```

The docs over here: http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-route53-aliastarget.html definitely seem to say to me this would be a problem. If you hit up the ruby SDK guys let me know, I would be interested to see what they say. Not accepting an empty list definitely seems weird. I'm still surprised there is so much drift between standards in the different SDKs.

**Comment by swelljoe** [comment]
Got it sorted (but, it needs more code to make it handle every case of AliasTarget).

So, ResourceRecords is actually not needed in an AliasTarget, so I'm deleting any empty ones.

It _does_ require a correct HostedZoneId, which is going to be different depending on what we're doing...and a restored domain will need to either:
1. Get the hosted zone ID from the newly created one, if the old zone AliasTarget was pointing to something within the zone (rather than CloudFront or Load Balancer).
2. Set the hosted zone ID to either the magic one for CloudFront or the one provided by the Load Balancer (or S3, or EBS). This may be pullable from the backed up zone. I don't think these change, unless they are recreated.

So, the script needs to decide, "Do I get the host_zone_id from the newly restored zone, or from the zone records backup?" The problem is I don't see immediately how to automatically detect _which_ one of those behaviors we need to use.

One option is to compare to see if the backed up zoneinfo zone ID matches the one in the records backup. That would, I think, mean that we need to swap it for the new one. If it does not match, we need to use the one from the backed up records file, as it points to an outside resource that these backups don't interact with. I think I'll give that a shot, and see if it works for your use case.  At the moment, it's working for me for zones that have an AliasTarget that points to something within the zone itself...maybe I'll create an S3 resource or something and see if it works for that, which I think would cover all of the other use cases.

**Comment by gravyboat** [comment]
What a pain. Thanks for taking a look into this!

**Comment by gravyboat** [comment]
Hmm, just wanted to provide an update today. I tried to restore that example domain I provided (with the latest pull and a `bundle exec update` and running:  `bundle exec bin/r53z --restore zones/ domain.com` and I get `Invalid request`. Were you able to get around this? I double checked to see if that dropped it in and was just an output error, but it only had 2 records instead of 20.

Edit: Actually is the latest code pushed to the repo? Your last update was 4 days ago but the last commit message was 5.
